### PR TITLE
preserve original tensoriterator behavior when not explicitly promoting

### PR DESCRIPTION
**Pull request by nairbv**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28253 Compute correct strides after type promotion
* **#28231 preserve original tensoriterator behavior when not explicitly promoting**

Summary:

Fix: https://github.com/pytorch/pytorch/issues/28010

A mixed-type index assignment that would have been an error in 1.2 was unintentionally made possible (with incorrect results) in 1.3. This PR restores the original behavior.

This is BC-breaking because:
```
        a = torch.ones(5, 2, dtype=torch.double)
        b = torch.zeros(5, dtype=torch.int)
        a[:, [1]] = b.unsqueeze(-1)
```
now raises an error (as in 1.2) whereas it did not in 1.3.

Differential Revision: [D18049637](https://our.internmc.facebook.com/intern/diff/D18049637)

### CODE REVIEW

File: aten/src/ATen/native/TensorIterator.cpp [hunk 1]
```diff
@@ -198,24 +200,24 @@ void TensorIterator::compute_types() {
           }
         }
       }
+    }
+  }
```
Comment (nairbv) [inline aten/src/ATen/native/TensorIterator.cpp]: before type promotion was added, this was two separate loops. Combining them presumes we only need to `validate_dtype` below when `missing_dtypes || compute_common_dtype`, but that assumption was incorrect.

File: aten/src/ATen/native/TensorIterator.cpp [hunk 2]
```diff
@@ -102,19 +102,21 @@ std::tuple<Device, ScalarType> TensorIterator::compute_common_type() {
...
+    if (op.is_output && check_output) {
       if (!canCast(common_dtype, op.tensor.scalar_type()))
       {
         AT_ERROR("result type ", common_dtype,
           " can't be cast to the desired output type ",
           op.tensor.scalar_type());
       }
     }
-    if (ninputs < 2 && op.dtype != op.tensor.scalar_type()) {
+    if (!promoting && op.dtype != op.tensor.scalar_type()) {
```
Comment (nairbv) [inline aten/src/ATen/native/TensorIterator.cpp]: Need to be more explicit on which tensor iterators use promotion. Initial implementation only impacted binary ops which all had two tensor arguments, but indexing ops have an arbitrary number of tensor arguments and now use tensor iterator.

**Comment by ifedan** [comment]
In general looks good, but I think it would be nice to add some tests to tensor_iterator_test.cpp

**Review by ifedan — APPROVED** [review]

**Pull request closed without merging**